website/database.py

```diff
@@ -608,6 +611,52 @@ def batch_get_adventures(self, adventure_ids):
         keys = {id: {"id": id} for id in adventure_ids}
         return ADVENTURES.batch_get(keys) if keys else {}
 
+    def get_public_adventures_filters(self):
+        filters = {}
+        for record in PUBLIC_ADVENTURES_FILTERS.scan():
+            if record.get("field") == "tag":
+                filters["tag"] = filters.get("tag", [])
+                filters["tag"].append(record.get("value"))
+            else:
+                lang, level = record.get("value").split("#")
+                filters["lang"] = filters.get("lang", [])
+                filters["lang"].append(lang)
+                filters["level"] = filters.get("level", [])
+                filters["level"].append(level)
+        return filters
+
+    def update_public_adventure_filters_indexes(self, adventure):
+        """This function adds/updates filters and indexes that we use to retrieve public adventures."""
+        if not adventure.get("public"):
+            return
+        id = adventure.get("id")
+        date = adventure.get("date")
+        lang_value = adventure.get("language", "en")
+        PUBLIC_ADVENTURES_INDEXES.put({"field_value": f"lang_{lang_value}", "date_adventure_id": f"{date}_{id}"})
+
+        levels = adventure.get("levels")
+        if not levels:
+            levels = [adventure.get("level")]
+        for level in levels:
+            PUBLIC_ADVENTURES_INDEXES.put({"field_value": f"level_{level}", "date_adventure_id": f"{date}_{id}"})
+            value = f"{lang_value}#{level}"
+            PUBLIC_ADVENTURES_FILTERS.put({"field": "lang#level", "value": value})
+        tags = adventure.get("tags", [])
+        for tag in tags:
+            PUBLIC_ADVENTURES_INDEXES.put({"field_value": f"tag#{tag}", "date_adventure_id": f"{date}_{id}"})
+            PUBLIC_ADVENTURES_FILTERS.put({"field": "tag", "value": tag})
+
+    def remove_public_adventure_filters_indexes(self, field, value, adventure):
+        PUBLIC_ADVENTURES_FILTERS.delete({"field": field, "value": value})
```

Comment:
The way we're storing information, you can't ever delete from `public_adventure_filters`.

Think of the following sequence of events:

* Make an adventure in Spanish public → `filters.put({ "field": "lang", "value": "es" })`  # Adds a new row
* Make another adventure in Spanish public → `filters.put({ "field": "lang", "value": "es" })`  # Replaces the row
* Make one of the Spanish adventures private → `filters.delete({ "field": "lang", "value": "es" })`  # Deletes the row

Oops! We now no longer know that there are any Spanish adventures!

Reply:
I also feel like `remove_public_daventure_filters_indexes` should be private, and be called from this class and this class alone. That way, callers don't have to be aware of this (and also can't forget!)

```
_remove_public_adventure_filters_indexes
```